File: packages/cli/hieudoanm/cobra.go/src/cmd/gh/coc/coc.go
package coc

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"

	"github.com/AlecAivazis/survey/v2"
	"github.com/hieudoanm/jack/src/cmd/gh/shared"
	requests "github.com/hieudoanm/jack/src/libs/requests"
	"github.com/spf13/cobra"
)

type codeOfConduct struct {
	Key  string `json:"key"`
	Name string `json:"name"`
	URL  string `json:"url"`
	Body string `json:"body"`
}

func fetchCOC(key, output string, fetch shared.FetchFunc) error {
	body, err := fetch("https://api.github.com/codes_of_conduct", requests.Options{
		Header: http.Header{
			"Accept":     {"application/json"},
			"User-Agent": {"hieudoanm-cli"},
		},
	})
	if err != nil {
		return fmt.Errorf("error fetching codes of conduct: %w", err)
	}

	var codes []codeOfConduct
	if err := json.Unmarshal(body, &codes); err != nil {
		return fmt.Errorf("error parsing response: %w", err)
	}

	selectedKey := key
	if selectedKey == "" {
		var keys []string
		for _, c := range codes {
			keys = append(keys, c.Key)
		}

		prompt := &survey.Select{
			Message: "Select a Code of Conduct:",
			Options: keys,
		}
		if err := survey.AskOne(prompt, &selectedKey); err != nil {
			return fmt.Errorf("selection cancelled: %w", err)
		}
	}

	url := fmt.Sprintf("https://api.github.com/codes_of_conduct/%s", selectedKey)
	detailBody, err := fetch(url, requests.Options{
		Header: http.Header{
			"Accept":     {"application/json"},
			"User-Agent": {"hieudoanm-cli"},
		},
	})
	if err != nil {
		return fmt.Errorf("error fetching code of conduct: %w", err)
	}

	var coc codeOfConduct
	if err := json.Unmarshal(detailBody, &coc); err != nil {
		return fmt.Errorf("error parsing response: %w", err)
	}

	file, err := os.Create(output)
	if err != nil {
		return fmt.Errorf("error creating file: %w", err)
	}
	defer file.Close()

	if _, err := file.WriteString(coc.Body); err != nil {
		return fmt.Errorf("error writing file: %w", err)
	}

	fmt.Printf("Written %s (%s) to %s\n", coc.Name, coc.Key, output)
	return nil
}

func NewCommand() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "coc",
		Short: "Fetch a GitHub Code of Conduct",
		Long: `Fetch and save a GitHub Code of Conduct to a file.

Fetches the list of available codes of conduct from the GitHub API,
prompts the user to select one (or uses --key), then writes the body
to a file.`,
		Example: `  gh coc
  gh coc --key citizen_code_of_conduct
  gh coc --key contributor_covenant -o COC.md`,
		RunE: func(cmd *cobra.Command, args []string) error {
			jsonOutput, _ := cmd.Flags().GetBool("json")
			key, _ := cmd.Flags().GetString("key")
			output, _ := cmd.Flags().GetString("output")

			if jsonOutput {
				if key == "" {
					return fmt.Errorf("--key is required for JSON output")
				}
				url := fmt.Sprintf("https://api.github.com/codes_of_conduct/%s", key)
				detailBody, err := shared.FetchFuncDefault(url, requests.Options{
					Header: http.Header{
						"Accept":     {"application/json"},
						"User-Agent": {"hieudoanm-cli"},
					},
				})
				if err != nil {
					return fmt.Errorf("error fetching code of conduct: %w", err)
				}
				var coc codeOfConduct
				if err := json.Unmarshal(detailBody, &coc); err != nil {
					return fmt.Errorf("error parsing response: %w", err)
				}
				out, _ := json.MarshalIndent(map[string]interface{}{
					"name": coc.Name,
					"body": coc.Body,
				}, "", "  ")
				fmt.Println(string(out))
				return nil
			}

			return fetchCOC(key, output, shared.FetchFuncDefault)
		},
	}

	cmd.Flags().String("key", "", "Code of Conduct key (skip prompt)")
	cmd.Flags().StringP("output", "o", "CODE_OF_CONDUCT", "Output file path")
	return cmd
}

File: packages/cli/hieudoanm/clap.rs/src/cmd/gh/coc.rs
use anyhow::Context;
use serde::Deserialize;

#[derive(Debug, Deserialize)]
struct CodeOfConduct {
    key: String,
    name: String,
    url: String,
    body: String,
}

fn fetch_cocs() -> anyhow::Result<Vec<CodeOfConduct>> {
    let client = reqwest::blocking::Client::new();
    let resp = client
        .get("https://api.github.com/codes_of_conduct")
        .header("Accept", "application/json")
        .header("User-Agent", "hieudoanm-cli")
        .send()
        .context("error fetching codes of conduct")?;
    let codes: Vec<CodeOfConduct> = resp.json().context("error parsing response")?;
    Ok(codes)
}

fn fetch_coc(key: &str) -> anyhow::Result<CodeOfConduct> {
    let url = format!("https://api.github.com/codes_of_conduct/{key}");
    let client = reqwest::blocking::Client::new();
    let resp = client
        .get(&url)
        .header("Accept", "application/json")
        .header("User-Agent", "hieudoanm-cli")
        .send()
        .context("error fetching code of conduct")?;
    let coc: CodeOfConduct = resp.json().context("error parsing response")?;
    Ok(coc)
}

#[derive(clap::Args)]
pub struct Args {
    #[arg(long = "key", help = "Code of Conduct key (skip prompt)")]
    pub key: Option<String>,
    #[arg(
        short = 'o',
        long = "output",
        default_value = "CODE_OF_CONDUCT",
        help = "Output file path"
    )]
    pub output: String,
}

pub fn command() -> clap::Command {
    clap::Command::new("coc")
        .about("Fetch a GitHub Code of Conduct")
        .arg(
            clap::Arg::new("key")
                .long("key")
                .help("Code of Conduct key (skip prompt)"),
        )
        .arg(
            clap::Arg::new("output")
                .short('o')
                .long("output")
                .help("Output file path")
                .default_value("CODE_OF_CONDUCT"),
        )
}

pub async fn run(matches: &Args) -> anyhow::Result<()> {
    let key = matches.key.as_ref();
    let output = &matches.output;

    let selected = if let Some(k) = key {
        k.clone()
    } else {
        let codes = fetch_cocs()?;
        let keys: Vec<&str> = codes.iter().map(|c| c.key.as_str()).collect();
        if keys.is_empty() {
            anyhow::bail!("no codes of conduct available");
        }
        println!("Available codes of conduct:");
        for (i, k) in keys.iter().enumerate() {
            println!("  {}. {k}", i + 1);
        }
        print!("Select by key: ");
        std::io::Write::flush(&mut std::io::stdout())?;
        let mut input = String::new();
        std::io::stdin().read_line(&mut input)?;
        input.trim().to_string()
    };

    let coc = fetch_coc(&selected)?;
    std::fs::write(output, &coc.body).context("error writing file")?;
    println!("Written {} ({}) to {output}", coc.name, coc.key);
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_command_definition() {
        let cmd = command();
        assert!(!cmd.get_name().is_empty());
    }
}
